stan/math/rev/core/arena_allocator.hpp

```diff
@@ -23,23 +30,87 @@ struct arena_allocator {
   T* allocate(std::size_t n) {
     return ChainableStack::instance_->memalloc_.alloc_array<T>(n);
   }
+  /**
+   * Returns the maximum theoretically possible value of n,
+   *  for which the call allocate(n, 0) could succeed.
+   */
+  static constexpr size_type max_size() noexcept {
+    return std::numeric_limits<size_type>::max() / sizeof(T);
+  }
 
   /**
-   * No-op. Memory is dealocated by caling `recover_memory()`.
+   * No-op. Memory is deallocated by calling `recover_memory()`.
    */
   void deallocate(T* /*p*/, std::size_t /*n*/) noexcept {}
 
+  /**
+   * Constructs an object of type T in allocated uninitialized storage pointed
+   * to by p, using placement-new
+   * @param p a pointer to type `T`
+   * @param val The value to use during construction.
+   */
+  static inline void construct(pointer p, const_reference val) {
+    new (static_cast<void*>(p)) T(val);
+  }
+  /**
+   * Constructs an object of type U in allocated uninitialized storage pointed
+   * to by p, using placement-new
+   * @tparam U The type to construct.
+   * @tparam Args Parameter pack of types used in constructor.
+   * @param p a pointer of type `U`
+   * @param args Parameter pack of objects used in constructor.
+   */
+  template <class U, class... Args>
+  void construct(U* p, Args&&... args) {
+    new (static_cast<void*>(p)) U(std::forward<Args>(args)...);
+  }
+
+  /**
+   * Use the allocator `a` to construct an object of type `U` in allocated
+   *  uninitialized storage pointed to by `p`, using placement-new.
+   *  This function is used by the standard library containers when inserting,
+   *  copying, or moving elements.
+   * @tparam U The type to construct.
+   * @tparam Args Parameter pack of types used in constructor.
+   * @param a An arena allocator that performs the construction.
+   * @param p a pointer of type `U`
+   * @param args Parameter pack of objects used in constructor.
+   */
+  template <class U, class... Args>
+  static void construct(arena_allocator<U>& a, U* p, Args&&... args) {
+    a.construct(p, args...);
+  }
+
+  /**
+   * No-op. If anything needs destroyed on the stack this happens
+   *  when calling `recover_memory()`.
+   */
+  static inline void destroy(pointer p) { (static_cast<T*>(p))->~T(); }
```

Comment:
@t4c1 quick sanity check, if we have a `std::vector<std::vector<var, arena_allocator> arena_allocator>` does the inner std vector need to have it's destructor called? Should we be using something like [`std::scoped_allocator_adaptor`](https://en.cppreference.com/w/cpp/memory/scoped_allocator_adaptor) for these?

Reply:
Nope. `std::vector<var, arena_allocator<var>>` never needs to be destructed, regardless of what kind of structure it is in.

Reply:
> Are we promised that the only dynamic memory inside of an std::vector is the data?

From https://en.cppreference.com/w/cpp/container/vector:
>std::vector (for T other than bool) meets the requirements of Container, AllocatorAwareContainer ...

From https://en.cppreference.com/w/cpp/named_req/AllocatorAwareContainer
> An AllocatorAwareContainer is a Container that holds an instance of an Allocator and uses that instance in all its member functions to allocate and deallocate memory and to construct and destroy objects in that memory

From https://en.cppreference.com/w/cpp/container/vector/~vector:
> Destructs the vector. The destructors of the elements are called and the used storage is deallocated.

So I think we are handling everything destructor should do.